proposals/0057-events.md

```diff
@@ -0,0 +1,192 @@
+---
+simd: '0057'
+title: Events
+authors:
+  - Maximilian Schneider (max@mango.markets)
+  - Nicholas Clarke (nicholasgclarke@mango.markets)
+category: Standard
+type: Core
+status: Draft
+created: 2023-06-13
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+Propose a new event logging mechanism that is reliable and efficient.
+Browser clients should be able to receive emitted events with minimal latency.
+3rd party indexers should allow to horizontally scale access to historic
+events.
+
+## Motivation
+
+Application developers need an easy way to reliably track the on-chain state
+of their programs. Common use-cases for this would be: executing a market order
+and reporting the filled price to the user. Collect all orders on a market and
+aggregate them to an hourly candles. A specialized API exposed to the runtime
+and RPC would allow dapps to reliably record historical data as well as RPC
+providers to serve it effectively to any web client.
+
+## Alternatives Considered
+
+Anchor events are a very popular solution, they pass base64 encoded binary rust
+structs as string to the `sol_log_` syscall. The issue with this process is
+that the sol_log_ destination buffer is truncated on many RPCs for complex
+transactions with a lot of logs. As the setting is an operator setting
+some of them choose to have very large log buffers, but not every RPC operator
+provides this service to their customers. There is a further issue with RPCs
+and log truncation in that you also need to ensure that the RPCs feeding the
+bigtable historical transactions also had a large log buffer otherwise you can
+experience truncation fetching logs older than the RPCs local history even when
+you're pointed at an RPC with a large log buffer. Another issue is that these
+logs are used to also record runtime information (compute units consumed) and
+plaintext debug message, hence become very difficult to parse, which is why
+many developers avoid structured event logging via anchor and just use
+formatted strings.
+
+A lot of developers would prefer a reliable and universally compatible solution
+so recently Anchor has added a new feature called cpi_events. This is a
+convention on how to call your own program via CPI and encode a binary string
+as instruction data. This can be indexed via the RPC Transaction response, as
+innerInstruction data is always persisted. It is more reliable than sol_log_
+but it effectively reduces runtime limits like maximum CPI call depth for the
+actual application. It also causeses unnecessary overhead in replay as a new
+execution context neeeds to be created, for something as trivial as a log
```

Comment:
This is an extremely valid criticism, but this should get resolved in Runtime V2, where VM's are only spun up once per transactions as opposed to once per instruction